### DISCUSSION TITLE
[NODRIVER] - AttributeError: 'NoneType' object has no attribute 'aclose'

### DISCUSSION THREAD

**Discussion by keisunl1** [ultrafunkamsterdam/undetected-chromedriver#2076] [Q&A]
Hi everyone,

Getting this error after pushing and pulling my project onto a new computer. Using Python 3.10.9

Traceback (most recent call last):
  File ****, line 1320, in <module>
    uc.loop().run_until_complete(main_script()) 
  File ****, line 90, in run_until_complete
    return f.result()
  File ****, line 201, in result
    raise self._exception.with_traceback(self._exception_tb)
  File ****, line 232, in __step
    result = coro.send(None)
  File ****, line 622, in main_script
    bot = await uc.start(no_sandbox=True)
  File ****, line 96, in start
    return await Browser.create(config)
  File ****, line 91, in create
    await instance.start()
  File ****, line 352, in start
    raise Exception(
Exception: 
                ---------------------
                Failed to connect to browser
                ---------------------
                One of the causes could be when you are running as root.
                In that case you need to pass no_sandbox=True 
                
Exception ignored in atexit callback: <function deconstruct_browser at 0x7fa21a5ba560>
Traceback (most recent call last):
  File ****, line 146, in deconstruct_browser
    _.stop()
  File ****, line 558, in stop
    asyncio.get_event_loop().create_task(self.connection.aclose())
AttributeError: 'NoneType' object has no attribute 'aclose'

I've read some older posts and usually had found that using "pkill -9 -f "Google Chrome" had worked when I encountered this crash on my old computer, but I'm always getting this error now with no workaround on my new computer. I've also seen someone speak about a user_data_dir being corrupted but I'm not sure where to find that or fix it. Any work around for this?

**Comment by keisunl1** [comment]
Some things I've tried as I've been debugging:
- Downgrading Chrome to v129
- Tried running with `asyncio.run()` instead of `uc.loop().run_until_complete()`
- Changing websocket vers
- Adding sleep for browser initialization

The chrome instance seems to start as soon as the code crashes as well (opens up a window after catching the Exception).

**Comment by mdmintz** [comment]
That's due to a change in https://github.com/python-websockets/websockets (a dependency of NoDriver).
> ``pip install websockets==13.1`` will fix that until NoDriver is patched.

----

Or, in the meantime, you can use my patched version of Undetected Chromedriver and NoDriver in https://github.com/seleniumbase/SeleniumBase  ([UC Mode](https://github.com/seleniumbase/SeleniumBase/blob/master/help_docs/uc_mode.md) and [CDP Mode](https://github.com/seleniumbase/SeleniumBase/blob/master/examples/cdp_mode/ReadMe.md))
Lots of updates, such as an ad-block option, proxy option, etc.

Example:

```python
from seleniumbase import SB

with SB(uc=True, test=True, ad_block=True) as sb:
    url = "https://www.browserscan.net/bot-detection"
    sb.activate_cdp_mode(url)
    sb.cdp.flash("Test Results")
    sb.cdp.assert_element('strong:contains("Normal")')
    sb.sleep(1)
    sb.cdp.save_screenshot("b_scan.png")
```

![b_scan](https://github.com/user-attachments/assets/faf81974-dd14-4116-9d9e-71869b3242c4)
